src/wallet/rpcnames.cpp

```diff
@@ -483,11 +523,81 @@ name_firstupdate ()
   if (name.size () > MAX_NAME_LENGTH)
     throw JSONRPCError (RPC_INVALID_PARAMETER, "the name is too long");
 
-  const valtype rand = ParseHexV (request.params[1], "rand");
-  if (rand.size () > 20)
-    throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+  const bool fixedRand = request.params.size () >= 2 && !request.params[1].isNull ();
+  const bool fixedTxid = request.params.size () >= 3 && !request.params[2].isNull ();
+  valtype rand(20);
+  if (fixedRand)
+    {
+      rand = ParseHexV (request.params[1], "rand");
+      if (rand.size () > 20)
+        throw JSONRPCError (RPC_INVALID_PARAMETER, "invalid rand value");
+    }
+
+  uint256 prevTxid = uint256::ZERO; // if it can't find a txid, force an error
+  if (fixedTxid)
+    {
+      prevTxid = ParseHashV (request.params[2], "txid");
+    }
+  else
+    {
+      // Code slightly duplicates name_scan, but not enough to be able to refactor.
+      /* Make sure the results are valid at least up to the most recent block
+         the user could have gotten from another RPC command prior to now.  */
+      pwallet->BlockUntilSyncedToCurrentChain ();
+
+      LOCK (cs_main);
+
+      for (const auto& item : pwallet->mapWallet)
+        {
+          const CWalletTx& tx = item.second;
+          if (!tx.tx->IsNamecoin ())
+            continue;
+
+          CScript output;
+          CNameScript nameOp;
+          int nOut = -1;
+          for (unsigned i = 0; i < tx.tx->vout.size (); ++i)
```

Comment:
Since we don't need the actual index of the output here (just that it exists and the operation itself later to check the hash), I think it would be nicer to use a C++11 for-each loop here.